source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,871 @@
+`Command Execution`_ section.
+
+See `Background connections use connectTimeoutMS as the timeout for
+handshake commands`_.
+
+Server Monitoring
+-----------------
+
+Drivers MUST NOT use ``timeoutMS`` for commands executed by the server
+monitoring and RTT calculation threads.
+
+See `Monitoring threads do not use timeoutMS`_.
+
+Cursors
+-------
+
+For operations that create cursors, ``timeoutMS`` can either cap the lifetime
+of the cursor or be applied separately to the original operation and all
+``next`` calls. To support both of these use cases, these operations MUST
+support a ``timeoutMode`` option. This option is an enum with possible values
+``CURSOR_LIFETIME`` and ``ITERATION``. The default value depends on the type
+of cursor being created. Drivers MUST error if ``timeoutMode`` is set and
+``timeoutMS`` is not.
+
+When applying the ``timeoutMS`` option to ``next`` calls on cursors, drivers
+MUST ensure it applies to the entire call, not individual commands. For
+drivers that send ``getMore`` requests in a loop when iterating tailable
+cursors, the timeout MUST apply to the totality of all ``getMore``’s, not to
+each one individually. If a resume is required for a ``next`` call on a
+change stream, the timeout MUST apply to the entirety of the initial
+``getMore`` and all commands sent as part of the resume attempt.
+
+For ``close`` methods, drivers MUST allow ``timeoutMS`` to be overridden if
+doing so is possible in the language. If ``timeoutMS`` is explicitly
+overridden, it MUST be honored. Otherwise, if ``timeoutMS`` was applied to
+the operation that created the cursor, drivers MUST use ``min(timeoutMS,
+connectTimeoutMS)`` as the ``timeoutMS`` value for the ``killCursors``
+command if one is required. This applies to explicit ``close`` methods that
+can be called by users as well as implicit destructors that are automatically
+invoked when exiting resource blocks.
+
+See `Cursor close() methods use connectTimeoutMS as timeoutMS`_.
+
+Non-tailable Cursors
+~~~~~~~~~~~~~~~~~~~~
+
+For non-tailable cursors, the default value of ``timeoutMode`` is
+``CURSOR_LIFETIME``. If ``timeoutMS`` is set, drivers MUST apply it to the
+original operation and the lifetime of the created cursor. For example, if a
+``find`` is executed at time ``T``, the ``find`` and all ``getMore``’s on the
+cursor must finish by time ``T + timeoutMS``. When executing ``next`` calls
+on the cursor, drivers MUST use the remaining timeout as the ``timeoutMS``
+value for the operation but MUST NOT append a ``maxTimeMS`` field to
+``getMore`` commands. If there are documents remaining in a previously
+retrieved batch, the ``next`` method MUST return them even if the timeout has
+expired and MUST only return a timeout error if a ``getMore`` is required.
+
+If ``timeoutMode`` is set to ``ITERATION``, drivers MUST honor the
+``timeoutMS`` option for the initial operation but MUST NOT append a
+``maxTimeMS`` field to the command sent to the server. The original
```

Comment:
Any concern about issues with not setting `maxTimeMS` in this mode?  I understand why we can't, but seems like it could cause issues server-side

Reply:
Based on the conversations we've had with the server teams, we know that recent versions of the server (4.2+ IIRC) will cancel in-progress work for an operation once the socket is disconnected as long as the operation doesn't change the collection state (i.e. it will for reads but not for writes). Sounds like it's not an issue for read-only commands like `find` and some aggregations, but it might be an issue for $out/$merge aggregations?

Reply:
We shouldn't be iterating a cursor for $out/$merge, so is that relevant?

Reply:
If those cursors are guaranteed to be empty, my previous point can be ignored. I don't think there's any concerns because this is already how `socketTimeoutMS` works if the application doesn't also set the `maxTimeMS` option on these operations.

Reply:
Technically there would be no cursor, at least server side.

Reply:
Then I don't think there's any risk here given a sufficiently new server version. Can we resolve this or do you think it'd be a good idea to follow up with the server team to confirm our understanding first?

Reply:
Addressed in [this commit](https://github.com/mongodb/specifications/pull/881/commits/6bb5ce85e304e16032b0a6e37083c1189e339031). Drivers will now add `maxTimeMS` to aggregates with $out/$merge even if `timeoutMode=ITERATION`.